prawns/balances/src/processor.ts

```diff
@@ -15,15 +24,198 @@ if (!supportedNetworks.includes(network)) {
   })
-  .addEventHandler('Balances.Transfer', balanceTransferHandler(network, 0))
-  .addEventHandler('Balances.BalanceSet', balanceSetHandler(network, 0))
-  .addEventHandler('Balances.Endowed', balanceEndowedHandler(network, 0))
-  .addEventHandler('Balances.Deposit', balanceDepositHandler(network, 0))
-  .addEventHandler('Treasury.Awarded', treasuryAwardedHandler(network, 0))
-  .run();
+  .addEvent('Balances.Transfer', {
+    data: { event: { args: true } },
+  } as const)
+  .addEvent('Balances.BalanceSet', {
+    data: { event: { args: true } },
+  } as const)
+  .addEvent('Balances.Endowed', {
+    data: { event: { args: true } },
+  } as const)
+  .addEvent('Balances.Deposit', {
+    data: { event: { args: true } },
+  } as const)
+  .addEvent('Treasury.Awarded', {
+    data: { event: { args: true } },
+  } as const);
+
+// doing .run above breaks the types... bit odd
+processor.run(new TypeormDatabase(), processBatch);
+
+async function processBatch(ctx: Context): Promise<void> {
+  const models: Models = {
+    accounts: [],
+    changeEvents: [],
+    transfers: [],
+  };
+
+  // synchronously process all data in the batch, mutating models (above) as the batch's data cache
+  for (const block of ctx.blocks) {
+    for (const item of block.items) {
+      if (item.kind === 'call') {
+        // we have none, leaving for reference
+      } else if (item.kind === 'event') {
+        const blockNumber = BigInt(block.header.height);
+        const date = new Date(block.header.timestamp);
+        processEvent({
+          ctx,
+          item,
+          blockNumber,
+          date,
+          models,
+          network,
+        });
+      }
+    }
+  }
+
+  // save accounts first as they are FKeyed in the other models
+  const existingAccounts = await ctx.store.findBy<SubstrateBalanceAccount>(
+    SubstrateBalanceAccount,
+    { id: In(models.accounts.map((acc) => acc.id)) }
+  );
+  await ctx.store.save(mergeAccounts(models.accounts, existingAccounts));
+  await ctx.store.save([...models.transfers, ...models.changeEvents]);
+
+  ctx.log.info(
+    JSON.stringify(
+      {
+        batchStart: ctx.blocks[0].header.height,
+        batchEnd: ctx.blocks[ctx.blocks.length - 1].header.height,
+        accounts: models.accounts.length,
+        changeEvents: models.changeEvents.length,
+        transfers: models.transfers.length,
+      },
+      null,
+      2
+    )
+  );
+}
+
+async function processEvent(params: EventProcessorParams) {
+  switch (params.item.name) {
+    case 'Balances.Transfer': {
+      const { from, to, amount } = getBalancesTransferEvent(
+        params.ctx,
+        params.item.event,
+        network
+      );
+      handleBalanceChange({
+        eventType: SubstrateBalanceChangeEventType.BalancesTransfer,
+        processorParams: params,
+        eventParams: {
+          amount,
+          account: to,
+          from,
+        },
+      });
+      break;
+    }
```

Comment:
I just wish there was a better solution for this switch statement :D But don't have a better one